# Getting started with Azure ML Data Prep SDK
Copyright (c) Microsoft Corporation. All rights reserved.<br>
Licensed under the MIT License.


#### Note: Some features in this Notebook will _not_ work with the Private Preview version of the SDK; it assumes the Public Preview version.

Wonder how you can make the most of the Azure ML Data Prep SDK? In this "Getting Started" guide, we'll showcase a few highlights that make this SDK shine for big datasets where `pandas` and `dplyr` can fall short. Using the [Ford GoBike dataset](https://www.fordgobike.com/system-data) as an example, we'll cover how to build Dataflows that allow you to:

* [Read in data](#Read-in-data)
* [Get a profile of your data](#Get-data-profile)
* [Apply smart transforms by Microsoft Research](#Derive-by-example)
* [Filter quickly](#Filter-our-data)
* [Apply common data science transforms](#Transform-our-data)
* [Easily handle errors and assertions](#Assert-on-invalid-data)
* [Prepare your dataset for export and machine learning](#Export-for-machine-learning)

In [1]:
from IPython.display import display
from os import path
from tempfile import mkdtemp

import pandas as pd
import azureml.dataprep as dprep

## Read in data

Azure ML Data Prep supports many different file reading formats (i.e. CSV, Excel, Parquet), and also offers the ability to infer column types automatically. 

In [2]:
gobike = dprep\
    .read_csv(
        path='https://dprepdata.blob.core.windows.net/demo/ford_gobike/2017-fordgobike-tripdata.csv',
        inference_arguments=dprep.InferenceArguments.current_culture()
    )
gobike.head(5)

duration_sec                       start_time  \
0       80110.0 2017-12-31 16:57:39.654000+00:00   
1       78800.0 2017-12-31 15:56:34.842000+00:00   
2       45768.0 2017-12-31 22:45:48.411000+00:00   
3       62172.0 2017-12-31 17:31:10.636000+00:00   
4       43603.0 2017-12-31 14:23:14.001000+00:00   

                          end_time  start_station_id  \
0 2018-01-01 15:12:50.245000+00:00              74.0   
1 2018-01-01 13:49:55.617000+00:00             284.0   
2 2018-01-01 11:28:36.883000+00:00             245.0   
3 2018-01-01 10:47:23.531000+00:00              60.0   
4 2018-01-01 02:29:57.571000+00:00             239.0   

                                  start_station_name  start_station_latitude  \
0                              Laguna St at Hayes St               37.776435   
1  Yerba Buena Center for the Arts (Howard St at ...               37.784872   
2                             Downtown Berkeley BART               37.870348   
3                               8th St at Ringold St               37.774520   
4                      Bancroft Way at Telegraph Ave               37.868813   

   start_station_longitude  end_station_id  \
0              -122.426244            43.0   
1              -122.400876            96.0   
2              -122.267764           245.0   
3              -122.409449             5.0   
4              -122.258764           247.0   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Dolores St at 15th St             37.766210   
2                             Downtown Berkeley BART             37.870348   
3       Powell St BART Station (Market St at 5th St)             37.783899   
4                          Fulton St at Bancroft Way             37.867789   

   end_station_longitude  bike_id   user_type  member_birth_year member_gender  
0            -122.415929     96.0    Customer             1987.0          Male  
1            -122.426614     88.0    Customer             1965.0        Female  
2            -122.267764   1094.0    Customer                NaN                
3            -122.408445   2831.0    Customer                NaN                
4            -122.265896   3167.0  Subscriber             1997.0        Female

In order to iterate more quickly, we can take a sample of our data. Later, we can then apply the same transformations to the entire dataset.

In [3]:
sampled_gobike = gobike.take_sample(probability=0.1, seed=5)

## Get data profile

Let's understand what our data looks like. Azure ML Data Prep facilitates this process by offering data profiles that help us glimpse into column types and column summary statistics.

In [4]:
gobike.get_profile()

ColumnProfile
    name: duration_sec
    type: FieldType.DECIMAL

    min: 61.0
    max: 86369.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 381.8421435321134
    median: 595.9837506906349
    upper_quartile: 938.5741138032683
    std: 3444.146451247386
    mean: 1099.009520877422

ColumnProfile
    name: start_time
    type: FieldType.DATE

    min: 2017-06-28 09:47:36.347000+00:00
    max: 2017-12-31 23:59:01.261000+00:00
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: end_time
    type: FieldType.DATE

    min: 2017-06-28 09:52:55.338000+00:00
    max: 2018-01-01 15:12:50.245000+00:00
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: start_station_id
    type: FieldType.DECIMAL

    min: 3.0
    max: 340.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 23.848148131600635
    median: 67.18427817406452
    upper_quartile: 139.42430180307275
    std: 86.08307797095921
    mean: 95.03424475658852

ColumnProfile
    name: start_station_name
    type: FieldType.STRING

    min: 10th Ave at E 15th St
    max: Yerba Buena Center for the Arts (Howard St at 3rd St)
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: start_station_latitude
    type: FieldType.DECIMAL

    min: 37.3172979
    max: 37.88022244590679
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 37.7735913559721
    median: 37.783211475877295
    upper_quartile: 37.79531236950411
    std: 0.08630496061661774
    mean: 37.771652603110894

ColumnProfile
    name: start_station_longitude
    type: FieldType.DECIMAL

    min: -122.44429260492325
    max: -121.8741186
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: -122.41170653070694
    median: -122.39875282257843
    upper_quartile: -122.39103429266093
    std: 0.10557344899193394
    mean: -122.36392726512949

ColumnProfile
    name: end_station_id
    type: FieldType.DECIMAL

    min: 3.0
    max: 340.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 22.702361193995444
    median: 65.22613324081779
    upper_quartile: 134.21987129021295
    std: 84.9694914863546
    mean: 92.18404079276426

ColumnProfile
    name: end_station_name
    type: FieldType.STRING

    min: 10th Ave at E 15th St
    max: Yerba Buena Center for the Arts (Howard St at 3rd St)
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: end_station_latitude
    type: FieldType.DECIMAL

    min: 37.3172979
    max: 37.88022244590679
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 37.774232065528906
    median: 37.78329810124021
    upper_quartile: 37.79557128475191
    std: 0.08622383487119635
    mean: 37.771843749644646

ColumnProfile
    name: end_station_longitude
    type: FieldType.DECIMAL

    min: -122.44429260492325
    max: -121.8741186
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: -122.41012752595213
    median: -122.39855511689811
    upper_quartile: -122.39096192032446
    std: 0.10512220222934929
    mean: -122.36323553679931

ColumnProfile
    name: bike_id
    type: FieldType.DECIMAL

    min: 10.0
    max: 3733.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 788.6785454424829
    median: 1726.652793720984
    upper_quartile: 2519.963581272433
    std: 971.3569593530214
    mean: 1672.533078699254

ColumnProfile
    name: user_type
    type: FieldType.STRING

    min: Customer
    max: Subscriber
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: member_birth_year
    type: FieldType.DECIMAL

    min: 1886.0
    max: 1999.0
    count: 519700.0
    missing_count: 66541.0
    error_count: 0.0

    lower_quartile: 1974.3341624985283
    median: 1982.80075

In [5]:
sampled_gobike.get_profile()

ColumnProfile
    name: duration_sec
    type: FieldType.DECIMAL

    min: 61.0
    max: 85864.0
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 381.0173265588649
    median: 596.0824682091602
    upper_quartile: 936.3990413401431
    std: 3527.1849383367376
    mean: 1102.2291284978571

ColumnProfile
    name: start_time
    type: FieldType.DATE

    min: 2017-06-28 10:51:23.182000+00:00
    max: 2017-12-31 23:55:09.686000+00:00
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: end_time
    type: FieldType.DATE

    min: 2017-06-28 11:01:39.557000+00:00
    max: 2018-01-01 15:12:50.245000+00:00
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: start_station_id
    type: FieldType.DECIMAL

    min: 3.0
    max: 340.0
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 23.82299260050619
    median: 66.81449005522046
    upper_quartile: 139.6790865298709
    std: 86.09232732608726
    mean: 94.87848340501073

ColumnProfile
    name: start_station_name
    type: FieldType.STRING

    min: 10th Ave at E 15th St
    max: Yerba Buena Center for the Arts (Howard St at 3rd St)
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: start_station_latitude
    type: FieldType.DECIMAL

    min: 37.3172979
    max: 37.88022244590679
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 37.773594346717786
    median: 37.78325255020885
    upper_quartile: 37.795362857566715
    std: 0.08626372544371842
    mean: 37.771708918993944

ColumnProfile
    name: start_station_longitude
    type: FieldType.DECIMAL

    min: -122.44429260492325
    max: -121.8741186
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: -122.41157512442906
    median: -122.39882719487981
    upper_quartile: -122.39096385593315
    std: 0.10559301820942323
    mean: -122.36375576045955

ColumnProfile
    name: end_station_id
    type: FieldType.DECIMAL

    min: 3.0
    max: 338.0
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 22.34742112221029
    median: 65.60893574407544
    upper_quartile: 135.08124174966116
    std: 85.09162990442911
    mean: 91.9201396254798

ColumnProfile
    name: end_station_name
    type: FieldType.STRING

    min: 10th Ave at E 15th St
    max: Yerba Buena Center for the Arts (Howard St at 3rd St)
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: end_station_latitude
    type: FieldType.DECIMAL

    min: 37.3184498
    max: 37.88022244590679
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 37.77450364194883
    median: 37.78358862499172
    upper_quartile: 37.79555394254664
    std: 0.08619152451969307
    mean: 37.77190111029278

ColumnProfile
    name: end_station_longitude
    type: FieldType.DECIMAL

    min: -122.44429260492325
    max: -121.8741186
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: -122.40967464398858
    median: -122.39857551157675
    upper_quartile: -122.39085540596203
    std: 0.10507512085392584
    mean: -122.3629776153239

ColumnProfile
    name: bike_id
    type: FieldType.DECIMAL

    min: 10.0
    max: 3733.0
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 795.8904240211187
    median: 1723.039443196501
    upper_quartile: 2524.901114053501
    std: 970.5058870359009
    mean: 1674.5133936319962

ColumnProfile
    name: user_type
    type: FieldType.STRING

    min: Customer
    max: Subscriber
    count: 51853.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: member_birth_year
    type: FieldType.DECIMAL

    min: 1900.0
    max: 1999.0
    count: 51853.0
    missing_count: 6577.0
    error_count: 0.0

    lower_quartile: 1974.2949238618335
    median: 1982.7223690704195
    upp

It appears that we have quite a few missing values in `member_birth_year`. We also immediately see that we have some empty strings in our `member_gender` column. With the data profiler, we can quickly do a sanity check on our dataset and see where we might need to start data cleaning.

## Derive by example

Azure ML Data Prep comes with additional "smart" transforms created by Microsoft Research. Here, we'll look at how you can derive a new column by providing examples of input-output pairs. Rather than explicitly using regular expressions to extract dates or hours from datetimes, we can provide examples for Azure ML Data Prep to learn what the pattern is. In fact, these smart transformations can also handle more complex derivations like inferring the day of the week from datetimes.

In [6]:
sgb_derived = sampled_gobike\
    .to_string(
        columns=['start_time', 'end_time']
    )\
    .derive_column_by_example(
        source_columns='start_time',
        new_column_name='date',
        example_data=[('2017-12-31 16:57:39.6540', '2017-12-31'), ('2017-12-31 16:57:39', '2017-12-31')]
    )\
    .derive_column_by_example(
        source_columns='start_time',
        new_column_name='hour',
        example_data=[('2017-12-31 16:57:39.6540', '16')]
    )\
    .derive_column_by_example(
        source_columns='start_time',
        new_column_name='wday',
        example_data=[('2017-12-31 16:57:39.6540', 'Sunday')]
    )

## Filter our data

Let's verify that our derivations are correct by doing a bit of spot-checking.

In [7]:
sgb_derived.filter(dprep.col('wday') != 'Sunday').head(5)

duration_sec                  start_time      wday hour        date  \
0        3456.0  2017-12-30 23:46:13.358000  Saturday   23  2017-12-30   
1         204.0  2017-12-30 23:31:38.904000  Saturday   23  2017-12-30   
2         743.0  2017-12-30 22:35:13.114000  Saturday   22  2017-12-30   
3         328.0  2017-12-30 22:19:28.760000  Saturday   22  2017-12-30   
4         260.0  2017-12-30 21:22:40.116000  Saturday   21  2017-12-30   

                     end_time  start_station_id  \
0  2017-12-31 00:43:49.469000              75.0   
1  2017-12-30 23:35:03.121000              84.0   
2  2017-12-30 22:47:36.356000             285.0   
3  2017-12-30 22:24:57.489000               5.0   
4  2017-12-30 21:27:00.885000             277.0   

                             start_station_name  start_station_latitude  \
0                      Market St at Franklin St               37.773793   
1                                   Duboce Park               37.769200   
2                    Webster St at O'Farrell St               37.783521   
3  Powell St BART Station (Market St at 5th St)               37.783899   
4                     Morrison Ave at Julian St               37.333658   

   start_station_longitude  end_station_id          end_station_name  \
0              -122.421239            75.0  Market St at Franklin St   
1              -122.433812           107.0     17th St at Dolores St   
2              -122.431158            97.0     14th St at Mission St   
3              -122.408445            64.0      5th St at Brannan St   
4              -121.908586           278.0    The Alameda at Bush St   

   end_station_latitude  end_station_longitude  bike_id   user_type  \
0             37.773793            -122.421239   1642.0  Subscriber   
1             37.763015            -122.426497   2201.0  Subscriber   
2             37.768265            -122.420110   1628.0  Subscriber   
3             37.776754            -122.399018   2806.0  Subscriber   
4             37.331932            -121.904888    465.0  Subscriber   

   member_birth_year member_gender  
0             1972.0          Male  
1             1965.0          Male  
2             1993.0          Male  
3             1986.0          Male  
4             1991.0          Male

We can also filter on other column types; let's take a peek at rides that lasted over 5 hours.

In [8]:
sgb_derived.filter(dprep.col('duration_sec') > (60 * 60 * 5)).head(5)

duration_sec                  start_time       wday hour        date  \
0       80110.0  2017-12-31 16:57:39.654000     Sunday   16  2017-12-31   
1       22587.0  2017-12-31 13:51:04.538000     Sunday   13  2017-12-31   
2       18067.0  2017-12-30 04:20:13.938000   Saturday   04  2017-12-30   
3       54550.0  2017-12-29 10:02:38.086000     Friday   10  2017-12-29   
4       63627.0  2017-12-27 19:12:42.794000  Wednesday   19  2017-12-27   

                     end_time  start_station_id  \
0  2018-01-01 15:12:50.245000              74.0   
1  2017-12-31 20:07:32.139000             307.0   
2  2017-12-30 09:21:21.628000              70.0   
3  2017-12-30 01:11:48.539000              21.0   
4  2017-12-28 12:53:10.649000             249.0   

                                 start_station_name  start_station_latitude  \
0                             Laguna St at Hayes St               37.776435   
1                                        SAP Center               37.332692   
2                            Central Ave at Fell St               37.773311   
3  Montgomery St BART Station (Market St at 2nd St)               37.789625   
4                         Russell St at College Ave               37.858473   

   start_station_longitude  end_station_id  \
0              -122.426244            43.0   
1              -121.900084           307.0   
2              -122.444293            43.0   
3              -122.400811            84.0   
4              -122.253253           244.0   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                                         SAP Center             37.332692   
2  San Francisco Public Library (Grove St at Hyde...             37.778768   
3                                        Duboce Park             37.769200   
4                         Shattuck Ave at Hearst Ave             37.873792   

   end_station_longitude  bike_id user_type member_birth_year member_gender  
0            -122.415929     96.0  Customer              1987          Male  
1            -121.900084   1443.0  Customer              None                
2            -122.415929   1928.0  Customer              None                
3            -122.433812    209.0  Customer              None                
4            -122.268618   1804.0  Customer              1988          Male

## Transform our data

In addition to "smart" transformations, Azure ML Data Prep also supports many common data science transforms familiar to other industry-standard data science libraries. Here, we'll explore the ability to `summarize` and `replace`. We'll also get to use `join` when we handle assertions.

#### Summarize


In [9]:
sgb_summary = sgb_derived\
    .summarize(
        summary_columns=[
            dprep\
                .SummaryColumnsValue(
                    column_id='duration_sec', 
                    summary_column_name='duration_sec_mean', 
                    summary_function=dprep.SummaryFunction.MEAN
                )
        ],
        group_by_columns=['date']
    )
sgb_summary.head(5)

date  duration_sec_mean
0  2017-12-31        1982.801418
1  2017-12-30        1203.766423
2  2017-12-29        1287.324841
3  2017-12-28         835.146465
4  2017-12-27        1658.735955

Azure Data Prep also makes it easy to append this output of `summarize` to the original table based on the grouping variable. 

In [10]:
sgb_appended = sgb_derived\
    .summarize(
        summary_columns=[
            dprep\
                .SummaryColumnsValue(
                    column_id='duration_sec', 
                    summary_column_name='duration_sec_mean', 
                    summary_function=dprep.SummaryFunction.MEAN
                )
        ],
        group_by_columns=['date'],
        join_back=True
    )
sgb_appended.head(5)

duration_sec                  start_time    wday hour        date  \
0       80110.0  2017-12-31 16:57:39.654000  Sunday   16  2017-12-31   
1        3292.0  2017-12-31 23:46:32.403000  Sunday   23  2017-12-31   
2        1397.0  2017-12-31 23:55:09.686000  Sunday   23  2017-12-31   
3         422.0  2017-12-31 23:54:25.337000  Sunday   23  2017-12-31   
4        1130.0  2017-12-31 23:36:16.069000  Sunday   23  2017-12-31   

                     end_time  start_station_id  \
0  2018-01-01 15:12:50.245000              74.0   
1  2018-01-01 00:41:24.605000             284.0   
2  2018-01-01 00:18:26.721000              78.0   
3  2018-01-01 00:01:27.354000             139.0   
4  2017-12-31 23:55:06.096000              66.0   

                                  start_station_name  start_station_latitude  \
0                              Laguna St at Hayes St               37.776435   
1  Yerba Buena Center for the Arts (Howard St at ...               37.784872   
2                                Folsom St at 9th St               37.773717   
3           Garfield Square (25th St at Harrison St)               37.751017   
4                              3rd St at Townsend St               37.778742   

   start_station_longitude  end_station_id  \
0              -122.426244            43.0   
1              -122.400876            22.0   
2              -122.411647            15.0   
3              -122.411901            99.0   
4              -122.392741            23.0   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Howard St at Beale St             37.789756   
2  San Francisco Ferry Building (Harry Bridges Pl...             37.795392   
3                               Folsom St at 15th St             37.767037   
4                      The Embarcadero at Steuart St             37.791464   

   end_station_longitude  bike_id   user_type member_birth_year member_gender  \
0            -122.415929     96.0    Customer              1987          Male   
1            -122.394643   3058.0    Customer              None                 
2            -122.394203   1667.0    Customer              None                 
3            -122.415442   2415.0  Subscriber              1985          Male   
4            -122.391034   2721.0    Customer              None                 

   duration_sec_mean  
0        1982.801418  
1        1982.801418  
2        1982.801418  
3        1982.801418  
4        1982.801418

#### Replace

Recall that our `member_gender` column had empty strings that stood in place of `None`. Let's use our `replace` function to properly recode them as `None`s.

In [11]:
sgb_replaced = sampled_gobike.replace_na(columns=['member_gender'])
sgb_replaced.head(5)

duration_sec                       start_time  \
0       80110.0 2017-12-31 16:57:39.654000+00:00   
1        3292.0 2017-12-31 23:46:32.403000+00:00   
2        1397.0 2017-12-31 23:55:09.686000+00:00   
3         422.0 2017-12-31 23:54:25.337000+00:00   
4        1130.0 2017-12-31 23:36:16.069000+00:00   

                          end_time  start_station_id  \
0 2018-01-01 15:12:50.245000+00:00              74.0   
1 2018-01-01 00:41:24.605000+00:00             284.0   
2 2018-01-01 00:18:26.721000+00:00              78.0   
3 2018-01-01 00:01:27.354000+00:00             139.0   
4 2017-12-31 23:55:06.096000+00:00              66.0   

                                  start_station_name  start_station_latitude  \
0                              Laguna St at Hayes St               37.776435   
1  Yerba Buena Center for the Arts (Howard St at ...               37.784872   
2                                Folsom St at 9th St               37.773717   
3           Garfield Square (25th St at Harrison St)               37.751017   
4                              3rd St at Townsend St               37.778742   

   start_station_longitude  end_station_id  \
0              -122.426244            43.0   
1              -122.400876            22.0   
2              -122.411647            15.0   
3              -122.411901            99.0   
4              -122.392741            23.0   

                                    end_station_name  end_station_latitude  \
0  San Francisco Public Library (Grove St at Hyde...             37.778768   
1                              Howard St at Beale St             37.789756   
2  San Francisco Ferry Building (Harry Bridges Pl...             37.795392   
3                               Folsom St at 15th St             37.767037   
4                      The Embarcadero at Steuart St             37.791464   

   end_station_longitude  bike_id   user_type member_birth_year member_gender  
0            -122.415929     96.0    Customer              1987          Male  
1            -122.394643   3058.0    Customer              None          None  
2            -122.394203   1667.0    Customer              None          None  
3            -122.415442   2415.0  Subscriber              1985          Male  
4            -122.391034   2721.0    Customer              None          None

## Assert on invalid data 

Azure ML Data Prep helps prevent broken pipelines and safeguard against bad data by supporting assertions. In our case, we'll create assertions to handle potentially erroneous `member_birth_year` values. The oldest person on record is no more than 130 years old, so birth year listed as before 1900 is wrong. Though our `sampled_gobike` dataset doesn't have any issues, we would fail on the full `gobike` dataset if we made that assumption. However, Azure ML Data Prep allows us to handle these gracefully with assertions.

In [12]:
gb_asserted = gobike\
    .assert_value(
        columns='member_birth_year', 
        expression=dprep.f_or(dprep.value.is_null(), dprep.value >= 1900),
        error_code='InvalidDate'
    )
gb_asserted.get_profile()

ColumnProfile
    name: duration_sec
    type: FieldType.DECIMAL

    min: 61.0
    max: 86369.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 381.8421435321134
    median: 595.9837506906349
    upper_quartile: 938.5741138032683
    std: 3444.146451247386
    mean: 1099.009520877422

ColumnProfile
    name: start_time
    type: FieldType.DATE

    min: 2017-06-28 09:47:36.347000+00:00
    max: 2017-12-31 23:59:01.261000+00:00
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: end_time
    type: FieldType.DATE

    min: 2017-06-28 09:52:55.338000+00:00
    max: 2018-01-01 15:12:50.245000+00:00
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: start_station_id
    type: FieldType.DECIMAL

    min: 3.0
    max: 340.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 23.848148131600635
    median: 67.18427817406452
    upper_quartile: 139.42430180307275
    std: 86.08307797095921
    mean: 95.03424475658852

ColumnProfile
    name: start_station_name
    type: FieldType.STRING

    min: 10th Ave at E 15th St
    max: Yerba Buena Center for the Arts (Howard St at 3rd St)
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: start_station_latitude
    type: FieldType.DECIMAL

    min: 37.3172979
    max: 37.88022244590679
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 37.7735913559721
    median: 37.783211475877295
    upper_quartile: 37.79531236950411
    std: 0.08630496061661774
    mean: 37.771652603110894

ColumnProfile
    name: start_station_longitude
    type: FieldType.DECIMAL

    min: -122.44429260492325
    max: -121.8741186
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: -122.41170653070694
    median: -122.39875282257843
    upper_quartile: -122.39103429266093
    std: 0.10557344899193394
    mean: -122.36392726512949

ColumnProfile
    name: end_station_id
    type: FieldType.DECIMAL

    min: 3.0
    max: 340.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 22.702361193995444
    median: 65.22613324081779
    upper_quartile: 134.21987129021295
    std: 84.9694914863546
    mean: 92.18404079276426

ColumnProfile
    name: end_station_name
    type: FieldType.STRING

    min: 10th Ave at E 15th St
    max: Yerba Buena Center for the Arts (Howard St at 3rd St)
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: end_station_latitude
    type: FieldType.DECIMAL

    min: 37.3172979
    max: 37.88022244590679
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 37.774232065528906
    median: 37.78329810124021
    upper_quartile: 37.79557128475191
    std: 0.08622383487119635
    mean: 37.771843749644646

ColumnProfile
    name: end_station_longitude
    type: FieldType.DECIMAL

    min: -122.44429260492325
    max: -121.8741186
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: -122.41012752595213
    median: -122.39855511689811
    upper_quartile: -122.39096192032446
    std: 0.10512220222934929
    mean: -122.36323553679931

ColumnProfile
    name: bike_id
    type: FieldType.DECIMAL

    min: 10.0
    max: 3733.0
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

    lower_quartile: 788.6785454424829
    median: 1726.652793720984
    upper_quartile: 2519.963581272433
    std: 971.3569593530214
    mean: 1672.533078699254

ColumnProfile
    name: user_type
    type: FieldType.STRING

    min: Customer
    max: Subscriber
    count: 519700.0
    missing_count: 0.0
    error_count: 0.0

ColumnProfile
    name: member_birth_year
    type: FieldType.DECIMAL

    min: 1900.0
    max: 1999.0
    count: 519700.0
    missing_count: 66541.0
    error_count: 2.0

    lower_quartile: 1974.3343079021402
    median: 1982.80080

Now, we can filter to see what caused the 2 errors above:

In [13]:
gb_errors = gb_asserted.filter(dprep.col('member_birth_year').is_error())
gb_errors.head(5)

duration_sec                       start_time  \
0        2546.0 2017-08-19 17:47:32.110000+00:00   
1        1767.0 2017-08-19 13:20:02.170000+00:00   

                          end_time  start_station_id  \
0 2017-08-19 18:29:58.825000+00:00             197.0   
1 2017-08-19 13:49:29.735000+00:00             235.0   

            start_station_name  start_station_latitude  \
0  El Embarcadero at Grand Ave               37.808848   
1          Union St at 10th St               37.807239   

   start_station_longitude  end_station_id             end_station_name  \
0               -122.24968           172.0      College Ave at Taft Ave   
1               -122.28937           197.0  El Embarcadero at Grand Ave   

   end_station_latitude  end_station_longitude  bike_id user_type  \
0             37.841800            -122.251535   1448.0  Customer   
1             37.808848            -122.249680     91.0  Customer   

                                   member_birth_year member_gender  
0  azureml.dataprep.native.DataPrepError("'Invali...          Male  
1  azureml.dataprep.native.DataPrepError("'Invali...          Male

#### Join
But what were the original values? Let's use `join` to figure out what the values were that caused our assert to throw an error. 

In [14]:
gb_errors.join(
    left_dataflow=gb_errors,
    right_dataflow=gobike,
    join_key_pairs=[
        ('duration_sec', 'duration_sec'),
        ('start_station_id', 'start_station_id'),
        ('bike_id', 'bike_id')
    ]
).head(5)

l_duration_sec                     l_start_time  \
0          2546.0 2017-08-19 17:47:32.110000+00:00   
1          1767.0 2017-08-19 13:20:02.170000+00:00   

                        l_end_time  l_start_station_id  \
0 2017-08-19 18:29:58.825000+00:00               197.0   
1 2017-08-19 13:49:29.735000+00:00               235.0   

          l_start_station_name  l_start_station_latitude  \
0  El Embarcadero at Grand Ave                 37.808848   
1          Union St at 10th St                 37.807239   

   l_start_station_longitude  l_end_station_id           l_end_station_name  \
0                 -122.24968             172.0      College Ave at Taft Ave   
1                 -122.28937             197.0  El Embarcadero at Grand Ave   

   l_end_station_latitude       ...        r_start_station_latitude  \
0               37.841800       ...                       37.808848   
1               37.808848       ...                       37.807239   

   r_start_station_longitude r_end_station_id           r_end_station_name  \
0                 -122.24968            172.0      College Ave at Taft Ave   
1                 -122.28937            197.0  El Embarcadero at Grand Ave   

  r_end_station_latitude  r_end_station_longitude r_bike_id r_user_type  \
0              37.841800              -122.251535    1448.0    Customer   
1              37.808848              -122.249680      91.0    Customer   

   r_member_birth_year r_member_gender  
0               1886.0            Male  
1               1886.0            Male  

[2 rows x 30 columns]

If we look at `r_member_birth_year`, we see that these people were listed as being born in 1886. That's impossible! Now that we've identified outliers and anomalies, we can appropriately clean our data however we like.

## Export for machine learning

One of the beautiful features of Azure ML Data Prep is that you only need to write your code once and choose whether to scale up or out; it takes care of figuring out how. To do so, you can export the `.dprep` file you've written tested on a smaller dataset, then run it with your larger dataset. Here, we show how you can export your new package. For a more detailed example on how to execute it on Spark, check out our [New York Taxicab scenario](https://github.com/Microsoft/PendletonDocs/blob/master/Scenarios/NYTaxiCab/01.new_york_taxi.ipynb).

In [ ]:
gobike = gobike.set_name(name="gobike")
package_path = path.join(mkdtemp(), "gobike.dprep")

print("Saving package to: {}".format(package_path))
package = dprep.Package(arg=gobike)
package.save(file_path=package_path)

## Want more information?

Congratulations on finishing your introduction to the Azure ML Data Prep SDK! If you'd like more detailed tutorials on how to construct machine learning datasets or dive deeper into all of its functionality, you can find more information in our detailed notebooks [here](https://github.com/Microsoft/PendletonDocs). There, we cover topics including how to:

* Cache your Dataflow to speed up your iterations
* Add your custom Python transforms
* Impute missing values
* Sample your data
* Reference and link between Dataflows
* Apply your Dataflow to a new, larger data source